### PR TITLE
Added CLI command to deploy selected services/modules in production

### PR DESCRIPTION
**Pull request by ThorodanBrom**
- In production, each service may be in different containers. To avoid
editing config to choose what to deploy, added the `--modules/-m` command
to the CLI
- comma separated module names passed as values or the `all` keyword
- If `all` is passed, or -m is not passed, all modules deployed
- Updated `deploy` method to take list of module names
- Overloaded `recursiveDeploy` to take list of module names
- Organised imports, formatting and added javadoc

### CODE REVIEW

**Review by abhi4578 — CHANGES_REQUESTED** [review]
This looks good! and I tested and it works as expected. maybe can we put in some where the command to do that as part of java docs.
Eg: for deploying Admin, token, registration, auditing verticle
``` 
java $$AUTH_JAVA_OPTS  -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar ./fatjar.jar  --host $$(hostname) -c configs/config.json -m iudx.aaa.server.admin.AdminVerticle,iudx.aaa.server.token.TokenVerticle,iudx.aaa.server.registration.RegistrationVerticle,iudx.aaa.server.auditing.AuditingVerticle 
```
for deploying all verticles:

```
java $$AUTH_JAVA_OPTS  -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar ./fatjar.jar  --host $$(hostname) -c configs/config.json -m all
```

 Also I have done some changes please review merge PR to your fork! @ThorodanBrom  https://github.com/ThorodanBrom/iudx-aaa-server-1/pull/1

**Comment by ThorodanBrom** [comment]
Hello Abhilash, thanks for checking the code. Regarding the documentation, we can add it to the Javadoc for `Deployer.java`. We can also add it to the README in a separate section called `JAR Based`. If you think this is fine, I'll make the changes. Is 
```
java -jar ./fatjar.jar  --host $$(hostname) -c configs/config.json -m all
```

enough, or would you prefer having the $$AUTH_JAVA_OPTS and the logging arg also. If the latter is needed, could you give some details on it? I don't think it's mentioned in the README anywhere.

**Comment by abhi4578** [comment]
> Hello Abhilash, thanks for checking the code. Regarding the documentation, we can add it to the Javadoc for `Deployer.java`. We can also add it to the README in a separate section called `JAR Based`. If you think this is fine, I'll make the changes. Is
> 
> ```
> java -jar ./fatjar.jar  --host $$(hostname) -c configs/config.json -m all
> ```
> 
> enough, or would you prefer having the $$AUTH_JAVA_OPTS and the logging arg also. If the latter is needed, could you give some details on it? I don't think it's mentioned in the README anywhere.

yeah we can have the whole thing, 
 - AUTH_JAVA_OPTS: - variable containing any java options to be passed to auth server java app. Eg:  max heap size - ```-Xmx4096m  ```
 -  ```-Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory```  option basically setting vertx logging java system property,  so that vertx components & libraries (rabbitmq client, vertx pgclient, eventloop logs, io.netty lib etc.) also logs using log4j2, [refer docs](https://vertx.io/docs/vertx-core/java/#_logging)

The command then is actually this  (the extra dollar is to just escape the ```$``` in docker-compose file)
```
java $AUTH_JAVA_OPTS  -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar ./fatjar.jar  --host $(hostname) -c configs/config.json -m iudx.aaa.server.admin.AdminVerticle,iudx.aaa.server.token.TokenVerticle,iudx.aaa.server.registration.RegistrationVerticle,iudx.aaa.server.auditing.AuditingVerticle 
```

**Comment by abhi4578** [comment]
where ```./fatjar``` is place holder which can be ( generated by using the command ```mvn clean package -Dmaven.test.skip=true```, generating the jars into ```target/``` folder)
```iudx.aaa.server-cluster-0.0.1-SNAPSHOT-fat.jar```  -  vertx clustered jar file, containing micrometer metrics
or ```iudx.aaa.server-dev-0.0.1-SNAPSHOT-fat.jar``` - vertx non-clustered and does not contain micrometer metrics

**Comment by abhi4578** [comment]
You can add the system property in pom exec plugin and can use when bringing up using ```mvn clean compile exec:java@aaa-server``` . The needed change is [here](https://github.com/abhi4578/iudx-aaa-server-1/commit/88c9321dae0527d402f177721a96d1150c21301f). Referred [this stackoverflow answer](https://stackoverflow.com/questions/3708846/how-to-pass-systemproperties-when-invoking-execjava-plugin-in-maven) for the change

**Comment by ThorodanBrom** [comment]
I have added the required documentation to the Javadoc as well as the README. 

Regarding the pom change for the logging, could you make a separate PR to this repo only (since it's not too related with this feature). I've omitted the logging argument from the docs because we won't need it even when running as a JAR once it's in the POM right?

If the doc changes are OK, the PR can be ready for review and merge.

**Comment by abhi4578** [comment]
> I have added the required documentation to the Javadoc as well as the README.
> 
> Regarding the pom change for the logging, could you make a separate PR to this repo only (since it's not too related with this feature). I've omitted the logging argument from the docs because we won't need it even when running as a JAR once it's in the POM right?
> 
> If the doc changes are OK, the PR can be ready for review and merge.

Hi bryan ,in jar based deployment we will have to yet specify the system property ```-Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory ``` in cmd line. The changes in the pom here, https://github.com/ThorodanBrom/iudx-aaa-server-1/pull/2  will be including that system property while deploying using ```mvn exec``` only.

File: README.md [hunk 1]
```diff
@@ -94,7 +94,51 @@ export LOG_LEVEL=INFO
...
+export AUTH_URL=http://<auth-domain-name>
+export LOG_LEVEL=INFO
+```
+3. Use maven to package the application as a JAR
+   `mvn clean package -Dmaven.test.skip=true`
+4. 2 JAR files would be generated in the `target/` directory
+    - `iudx.aaa.server-cluster-0.0.1-SNAPSHOT-fat.jar` - clustered vert.x containing micrometer metrics
+    - `iudx.aaa.server-dev-0.0.1-SNAPSHOT-fat.jar` - non-clustered vert.x and does not contain micrometer metrics
+
+#### Running the clustered JAR
```
Comment (abhi4578) [inline README.md:110]: Maybe we can add a point here. 
1. Need to setup the zookeeper : https://zookeeper.apache.org/doc/r3.3.3/zookeeperStarted.html
2. Edit the config file option "zookeeper" should point to  ip/domain the zookeeper is hosted. The aaa server assumes the default port 2181 for zookeeper related communication and 5701 port for hazelcast communication between different vertx instances.

File: README.md [hunk 2]
```diff
@@ -94,7 +94,51 @@ export LOG_LEVEL=INFO
...
+* --modules/-m : comma separated list of module names to deploy</li>
+
+ e.g. `java -jar target/iudx.aaa.server-cluster-0.0.1-SNAPSHOT-fat.jar --host $(hostname) -c configs/config.json -m iudx.aaa.server.admin.AdminVerticle,iudx.aaa.server.token.TokenVerticle ,iudx.aaa.server.registration.RegistrationVerticle,iudx.aaa.server.auditing.AuditingVerticle`
+
+Use the `--help/-h` argument for more information. You may additionally append an `AUTH_JAVA_OPTS` environment variable containing any Java options to pass to the application.
 
+e.g.
+```
+$ export AUTH_JAVA_OPTS="-Xms128m -Xmx512m"
+$ java AUTH_JAVA_OPTS -jar target/iudx.aaa.server-cluster-0.0.1-SNAPSHOT-fat.jar ...
```
Comment (abhi4578) [inline README.md:124]: AUTH_JAVA_OPTS should have a``` $``` in front.  It should be like this 
```java $AUTH_JAVA_OPTS -jar target/iudx.aaa.server-cluster-0.0.1-SNAPSHOT-fat.jar ...```
In docker-compose, extra ```$``` was to escape the ```$``` :)

**Comment by abhi4578** [comment]
Added the PR https://github.com/datakaveri/iudx-aaa-server/pull/110, Regarding the pom change for the logging.

**Comment by ThorodanBrom** [comment]
Hello Abhilash, I've added the suggestions. We'll need to see if it's possible to force the server to use log4j2 always, since it's how we want the logs to be formatted. It would be one less thing to worry about when deploying, and I do not think it is something we need to keep customizable (correct me if I'm wrong here).

I've kept the note about zookeeper quite brief for now. I think that going forward we may need to add details on why zookeeper or hazelcast is used and maybe why running in clustered mode is advantageous etc. This ideally should be part of our developer docs since it's common for all IUDX services. The README here can then link to the devdocs for an overview of clustered Vert.x + zookeeper/hazelcast

**Comment by abhi4578** [comment]
> 
So the java system property makes the vertx internal components to use log4j2 as backend. There is not customizing of log format , you can customise the log level and auth url using env variables

**Review by abhi4578 — APPROVED** [review]
LGTM

**Pull request merged**